### DISCUSSION TITLE
C Intellisense is not working on Remote SSH

### DISCUSSION THREAD

**Discussion by 0xm4s0ud** [microsoft/vscode-cpptools#9164] [Q&A] [answered]
Hi,
I have a problem with C/C++ Intellisense on remote SSH. it stucks on "Updating Intellisense" and does not work properly.
I almost have tried all of solutions provided in community. i am sure that problem is not related to `include path` and my `settings.json` is in default mode. also there is no non-ascii character in the remote path. so, below is the part of logs i extracted from intellisense log and i realized that the remote vscode cannot spawn intellisense process using the "libuv" library and "uv_pipe_bind()" function due to the permission constraints. (-13 refers to `EACCESS`)
I have tried to uninstall vscode server and remove directories and nothing happend.
it is worth mentioning that i tried to connect to another ubuntu server with exactly same repository, same os, same vscode and everything was working as expected and there was not any problem.
The remote host is ubuntu 14.04 x64

C/C++ intellisense log:

`loggingLevel: Debug`
`cpptools/didChangeCppProperties`
`Attempting to get defaults from C compiler in "compilerPath" property: '/usr/bin/gcc'`
`Compiler query command line: "/usr/bin/gcc" -std=c89 -Wp,-v -E -dM -x c -m64  /dev/null`
`terminating child process: 17882`
`Code browsing service initialized`
`Attempting to get defaults from C++ compiler in "compilerPath" property: '/usr/bin/gcc'`
`Compiler query command line: "/usr/bin/gcc" -std=gnu++98 -Wp,-v -E -dM -x c++ -m64  /dev/null`
`terminating child process: 17889`
 ` Folder: /usr/include/ will be indexed`
 ` Folder: /usr/lib/gcc/x86_64-linux-gnu/4.8/include/ will be indexed`
 ` Folder: /usr/lib/gcc/x86_64-linux-gnu/4.8/include-fixed/ will be indexed`
 ` Folder: /usr/local/include/ will be indexed`
 ` Folder: /home/[user]/[source] will be indexed`
`textDocument/didOpen: /home/[user]/[source]/[file.c]`
`Discovering files...`
`cpptools/textEditorSelectionChange`
`cpptools/activeDocumentChange: /home/[user]/[source]/[file.c]`
 ` Processing folder (recursive): /usr/include/`
`cpptools/getDocumentSymbols: /home/[user]/[source]/[file.c] (id: 2)`
`cpptools/getCodeActions: /home/[user]/[source]/[file.c] (id: 3)`
`cpptools/getDocumentSymbols`
`cpptools/getFoldingRanges: /home/[user]/[source]/[file.c] (id: 4)`
 ` Processing folder (recursive): /usr/lib/gcc/x86_64-linux-gnu/4.8/include/`
 ` Processing folder (recursive): /usr/lib/gcc/x86_64-linux-gnu/4.8/include-fixed/`
 ` Processing folder (recursive): /usr/local/include/`
`uv_pipe_bind failed with ret = -13`
`Failed to spawn IntelliSense process:`
`Shutting down IntelliSense server: /home/[user]/[source]/[file.c]`

**Comment by Colengms** [comment, marked as answer]
Hi @masoudjafari .  There is a "C++ IntelliSense" extension in the marketplace, but it is not developed by Microsoft.  If you have that extension installed, it would conflict with the "C/C++" extension from Microsoft.

Though, the log you have provided does appear to be from the C/C++ extension.  It appears to be failing to communicate with a child (IntelliSense) process that it started, due to a permission error (-13, EACCESS).   Since you mentioned that the same scenario works on another machine, the issue would seem to be related to the configuration of the machine it's failing on.  The IntelliSense process is a native process launched by `cpptools`, which is also a native process.  The permission error seems to have arisen when `cpptools` launches `cpptools-srv` and tries to open a communication pipe with it.  Is there anything unusual about the host configuration that might account for that?

Ubuntu 14.04 is somewhat dated.  It's possible an issue has arisen with that older version of Ubuntu for some reason.

**Reply by 0xm4s0ud** [reply]
There is only one intellisense in my vscode and it's Microsoft c/c++ extension suggested by vscode.
I had been working with this ubuntu machine remotely before this issue arised, so i think it is not related to ubuntu version.
Do you have any suggestion to check why `cpptools` failes to launch `cpptools-srv`? I have all of permissions on the both files.
There is a problem in remote machine that might related to the problem. when i run programs such as git or gcc to create or modify files, they change the owner of files to `root`. for example when i use gcc to compile a `.c` file it creates an object file with owner of `root`.
Thanks for help

**Reply by sean-mcmanus** [reply]
What version of the C/C++ extension are you using?

**Reply by 0xm4s0ud** [reply]
It is latest version 1.9.7
I have tested old versions and insiders and that could not help.
I think it's all about the permissions. would you suggest a way to test `cpptools` and `cpptools-srv` manually?
I can run both binary files in remote server but i dont see anything and tried to launch them with some options like `--help` or `-h` but nothing appears.

**Reply by Colengms** [reply]
Hi @masoudjafari .  The child process is actually launched successfully.  The issue occurs when opening a communication channel (pipe) to that process.

```
uv_pipe_bind failed with ret = -13
```

> I had been working with this ubuntu machine remotely before this issue arised

Do you recall what may have changed when it stopped working?  Perhaps an automatic update of the C/C++ extension introduced the issue?  Could you check if previous versions work properly?  If so, could you tell us what version the issue was introduced in?

**Reply by 0xm4s0ud** [reply]
> Do you recall what may have changed when it stopped working?

Actually i just repaired my root file system using `fsck` and this issue appeared right after that (for all users of remote machine).

> Could you check if previous versions work properly?

No, the previous versions have same problem.

**Reply by Colengms** [reply]
Under the hood, files are created associated with pipes.  It looks like we are attempting to creates pipes within the `/tmp` directory.  Is this directory properly writable by users?  Perhaps permission to it was modified when your file system was repaired (or caused by whatever led to needing to repair it)?

**Reply by 0xm4s0ud** [reply]
That's it @Colengms it works! I appreciate your help.
And as final question: the intellisense does not `include` some `header` files that are already exists in my project and my `includePath` contains only `"${workspaceFolder}/**"` which supposed to include all of `header` files recursively. do you have any suggestion?

**Reply by Colengms** [reply]
Hi @masoudjafari .  I'm assuming the header that can't be found is located somewhere under your workspace folder.  :)   Is this something you can reproduce with a simple example?  I'd suggest opening an issue for this, and providing the information requested in the issue submission form.  There is likely some detail in your configuration or environment that we may need to locate to determine what's happening there.
